isabelle/Core/Semantics.thy

```diff
@@ -726,34 +726,80 @@ and maxTimeCase :: "Case \<Rightarrow> int" where
-"geIfNone (Some x) y = (x \<ge> y)"
+"geIfNone (Some x) y = (x \<ge> y) "
 
 fun subIfSome :: "int option \<Rightarrow> int \<Rightarrow> int option" where
 "subIfSome None _ = None" |
 "subIfSome (Some x) y = Some (x - y)"
 
 
+text \<open>
+A TimeInterval (startTime, endTime) can be ambiguous wrt a When's timeout if startTime < timeout \<le> endTime
+
+\<close>
+
+text \<open>The \<^emph>\<open>calculateNonAmbiguousInterval\<close> function can help a user to calculate a TimeInterval that
+won't be ambiguous for the next \<^emph>\<open>n\<close> inputs of a contract.\<close>
+
+text \<open>The only Contract constructor that can yield a \<^term>\<open>TEAmbiguousTimeIntervalError\<close> is the \<^term>\<open>When\<close> case. 
+Computing a transaction of a contract that starts in a non quiescent state (\<^term>\<open>Let\<close>, \<^term>\<open>Pay\<close>, \<^term>\<open>If\<close>, \<^term>\<open>Assert\<close>) 
+can end in a \<^term>\<open>TEAmbiguousTimeIntervalError\<close> iff there is a \<^term>\<open>When\<close> case that makes it ambiguous.
+\<close>
+
+text \<open>A TimeInterval expressed as the tuple \<^term>\<open>(startTime \<times> endTime)\<close> can be ambiguous wrt a \<^term>\<open>When\<close>'s timeout
+iff \<^emph>\<open>startTime < timeout \<le> endTime\<close>
+\<close>
+
+text
+\<open> It parameters of \<^emph>\<open>calculateNonAmbiguousInterval\<close> are:
+\<^item> An optional number of Inputs to check. The number of inputs corresponds to the number of "When". 
+If None is passed, it means that we should check for transactions of any number of inputs.
+\<^item> A lower bound (normally the current time).
+\<^item> The Contract to check.
+
+The function returns an Optionally Bound Time Interval, as defined in the \<^emph>\<open>OptBoundTimeInterval.thy\<close> theory.
+In the \<^emph>\<open>TimeRange.thy\<close> theory we prove that computing a transaction with these bounds doesn't end with an ambiguous error. 
+\<close>
+\<comment> \<open>TODO: The intersectInterval can produce an invalid time interval, which would mean that not suitable TimeInterval was found\<close>
 fun calculateNonAmbiguousInterval :: "int option \<Rightarrow> POSIXTime \<Rightarrow> Contract \<Rightarrow> OptBoundTimeInterval"
-where
+  where
+\<comment> \<open>A Close contract can't be ambiguous, so an Unbounded interval is returned \<close>
 "calculateNonAmbiguousInterval _ _ Close = (Unbounded, Unbounded)" |
+\<comment> \<open>A Pay contract isn't ambiguous by itself, so we calculate for the continuation\<close>
 "calculateNonAmbiguousInterval n t (Pay _ _ _ _ c) = calculateNonAmbiguousInterval n t c" |
-"calculateNonAmbiguousInterval n t (If _ ct cf) = intersectInterval (calculateNonAmbiguousInterval n t ct)
+\<comment> \<open>If we branch, we intersect the result of both possibilites. \<close>
+\<comment> \<open>TODO: Note that not knowing which branch can be selected beforehand means that we return a smaller TimeInterval than
```

Comment:
well, the task is not explicitly said in imperative form but it does say: "Maybe we could improve this by using the actual Input instead of the number of inputs".

Changing this to `FIXME` in my local development of the semantics refactoring